app/controllers/add-preprint.js

```diff
@@ -1,87 +1,207 @@
+                    this.get('toast').info('Project deleted');
+                });
+                this.set('preprintNode', null);
+                // TODO: reset dropzone, since uploaded file has no project
+            }
+            resolve();
+        },
+        preUpload(ignore, dropzone, file) {
+            this.set('uploadFile', file);
+            return new Ember.RSVP.Promise(resolve => this.set('resolve', resolve));
+        },
+        startUpload(resolve) {
+            this.set('_url', `${this.get('preprintNode.files').findBy('name', 'osfstorage').get('links.upload')}?kind=file&name=${this.get('uploadFile.name')}`);
+            this.get('resolve')();
+            this.get('toast').info('File will upload in the background.');
+            resolve();
+        },
+        buildUrl() {
+            return this.get('_url');
+        },
+        success() {
+            this.get('toast').info('File uploaded!');
+        },
+        deleteSubject(key, array = key.split('.')) {
+            this.set(key, null);
+            // Delete key manually
+            switch (array.length) {
+                case 2:
+                    delete this[array[0]][array[1]];
+                    break;
+                case 3:
+                    delete this[array[0]][array[1]][array[2]];
+                    break;
+                case 4:
+                    delete this[array[0]][array[1]][array[2]][array[3]];
+                    break;
+                default:
+                    console.error('deletion not implemented');
+            }
+        },
+        deselectSubject([...args]) {
+            args = args.filter(arg => Ember.typeOf(arg) === 'string');
+            this.send('deleteSubject', `selected.${args.join('.')}`, ['selected', ...args]);
+            this.notifyPropertyChange('selected');
+            this.rerender();
+        },
+        selectSubject(...args) {
+            const process = (prev, cur, i, arr) => {
+                const selected = this.get(`selected.${prev}`);
+                if (!selected) {
+                    // Create necessary parent objects and newly selected object
+                    this.set(`selected.${prev}`, new Ember.Object());
+                } else if (i === 3 || i === args.length && args.length === this.get('path').length &&
+                    this.get('path').every((e, i) => e.name === args[i].name) &&
+                    Object.keys(selected).length === 0) {
+                    // Deselecting a subject: if subject is last item in args,
+                    // its children are showing, and no children are selected
+                    this.send('deleteSubject', `selected.${prev}`, ['selected', ...arr.splice(0, i)]);
+                    args.popObject();
+                }
+                return `${prev}.${cur}`;
+            };
+            // Process past length of array
+            [...args.map(arg => arg.name || arg), ''].reduce(process);
```

Comment:
This is code from a previous PR. We may not have time to fix it in this PR, but it's... pretty weird.